v3.8.2 issue with @types/jss

<!--- Provide a general summary of the issue in the Title above -->
Compile error:
```
ERROR in node_modules/@material-ui/core/styles/createGenerateClassName.d.ts(1,10):
TS2305: Module '"node_modules/@types/jss"' has no exported member 'GenerateId'.
ERROR in node_modules/@material-ui/core/styles/jssPreset.d.ts(1,10):
TS2724: Module '"node_modules/@types/jss"' has no exported member 'JssOptions'. Did you mean 'JSSOptions'?
ERROR in node_modules/@material-ui/core/styles/withStyles.d.ts(31,48):
TS2694: Namespace '"node_modules/@types/jss/index"' has no exported member 'StyleSheetFactoryOptions'.
```

<!--
    Thank you very much for contributing to Material-UI by creating an issue! ❤️
    To avoid duplicate issues we ask you to check off the following list.
-->

<!-- Checked checkbox should look like this: [x] -->
- [x] This is not a v0.x issue. <!-- (v0.x is no longer maintained) -->
- [x] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior 🤔
<!---
    Describe what should happen.
-->
No errors

## Current Behavior 😯
<!---
    Describe what happens instead of the expected behavior.
-->
Errors

## Steps to Reproduce 🕹
<!---
    Provide a link to a live example (you can use codesandbox.io) and an unambiguous set of steps to reproduce this bug.
    Include code to reproduce, if relevant (which it most likely is).

    This codesandbox.io template _may_ be a good starting point:
    https://codesandbox.io/s/github/mui-org/material-ui/tree/master/examples/create-react-app

    If you're using typescript a better starting point would be
    https://codesandbox.io/s/github/mui-org/material-ui/tree/master/examples/create-react-app-with-typescript

    If YOU DO NOT take time to provide a codesandbox.io reproduction, should the COMMUNITY take time to help you?

-->
Link:

1. update package.json to `"@material-ui/core": "3.8.1"`
2. rm -rf node_modules package-lock.json
3. npm i
4. npm start

## Context 🔦
<!---
    What are you trying to accomplish? How has this issue affected you?
    Providing context helps us come up with a solution that is most useful in the real world.
-->
I just upgraded from `3.7.1` to `3.8.1`. I have **no** entry for `jss` nor `@types/jss` in my `package.json`.

The only entries for `@types/jss` in my `package-lock.json`:
```
"@material-ui/core": {
  "version": "3.8.1",
  "resolved": "https://registry.npmjs.org/@material-ui/core/-/core-3.8.1.tgz",
  "integrity": "sha512-YhVanMT+DBaw8te6WbpU+RGm/wPYwT2FXlz0JGEexSOReDK3zVuyQJ7mTyDEE+HZQH6p1s6CTXXIH4CoivpogQ==",
  "requires": {
    "@babel/runtime": "7.2.0",
    "@material-ui/system": "^3.0.0-alpha.0",
    "@material-ui/utils": "^3.0.0-alpha.2",
    "@types/jss": "^9.5.6",
    "@types/react-transition-group": "^2.0.8",
    "brcast": "^3.0.1",
    ...



"@types/jss": {
  "version": "9.5.7",
  "resolved": "https://registry.npmjs.org/@types/jss/-/jss-9.5.7.tgz",
  "integrity": "sha512-OZimStu2QdDMtZ0h72JXqvLVbWUjXd5ZLk8vxLmfuC/nM1AabRyyGoxSufnzixrbpEcVcyy/JV5qeQu2JnjVZw==",
  "requires": {
    "csstype": "^2.0.0",
    "indefinite-observable": "^1.0.1"
  }
},
```

**EDIT** Happens with `@types/jss` `9.5.6` and `9.5.7`

## Your Environment 🌎
<!---
    Include as many relevant details about the environment with which you experienced the bug.
    If you encounter issues with typescript please include version and tsconfig.
-->

| Tech         | Version |
|--------------|---------|
| Material-UI  | v3.8.1  |
| React        | 16.7.0        |
| Browser      | Chrome 71.0        |
| TypeScript   | 3.2.2        |
| etc.         |         |

Hi @benneq @rosskevin @oliviertassinari 

I have been trying to update the material UI version used in my project from 1.4.3 to 3.8.1.

As discussed above, I had similar compiling issues with @types/jss.

I was able to resolve the compile errors with steps suggested by @benneq  (https://github.com/mui-org/material-ui/issues/14040#issuecomment-450654382)

Apart from this, I am facing another typescript error mentioned below which is coming up when I try and build the project.

error TS2370: A rest parameter must be of an array type.

Similar issue mentioned here :- https://github.com/goemen/react-material-ui-typescript/issues/1

Any help, appriciated thanks !!

--- a/docs/src/pages/demos/buttons/CustomizedButtons.js
+++ b/docs/src/pages/demos/buttons/CustomizedButtons.js
@@ -23,6 +23,7 @@ const styles = theme => ({
     fontSize: 16,
     padding: '6px 12px',
     border: '1px solid',
+    lineHeight: 1.5,
     backgroundColor: '#007bff',
     borderColor: '#007bff',
     fontFamily: [

--- a/packages/material-ui/src/Button/Button.js
+++ b/packages/material-ui/src/Button/Button.js
@@ -12,13 +12,11 @@ import { capitalize } from '../utils/helpers';
 export const styles = theme => ({
   /* Styles applied to the root element. */
   root: {
-    lineHeight: 1.3125, // To remove with v4.
+    lineHeight: 1.75, // To remove with v4.
     ...theme.typography.button,
     boxSizing: 'border-box',
     minWidth: 64,
-    minHeight: 36,
     padding: '6px 16px',
-    alignItems: 'center',
     borderRadius: theme.shape.borderRadius,
     color: theme.palette.text.primary,
     transition: theme.transitions.create(['background-color', 'box-shadow', 'border'], {
@@ -37,10 +35,6 @@ export const styles = theme => ({
     },
     '&$disabled': {
       color: theme.palette.action.disabled,
-    },
-    '&:before': {
-      content: '""',
-      height: 24, // Same height as desired min-height
     },
   },
   /* Styles applied to the span element that wraps the children. */
@@ -84,6 +78,7 @@ export const styles = theme => ({
   flatSecondary: {},
   /* Styles applied to the root element if `variant="outlined"`. */
   outlined: {
+    padding: '5px 16px',
     border: `1px solid ${
       theme.palette.type === 'light' ? 'rgba(0, 0, 0, 0.23)' : 'rgba(255, 255, 255, 0.23)'
     }`,
@@ -214,22 +209,12 @@ export const styles = theme => ({
   sizeSmall: {
     padding: '4px 8px',
     minWidth: 64,
-    minHeight: 31,
     fontSize: theme.typography.pxToRem(13),
-    '&:before': {
-      content: '""',
-      height: 23, // Same height as desired min-height
-    },
   },
   /* Styles applied to the root element if `size="large"`. */
   sizeLarge: {
     padding: '8px 24px',
-    minHeight: 42,
     fontSize: theme.typography.pxToRem(15),
-    '&:before': {
-      content: '""',
-      height: 26, // Same height as desired min-height
-    },
   },
   /* Styles applied to the root element if `fullWidth={true}`. */
   fullWidth: {

--- a/packages/material-ui/src/Button/Button.js
+++ b/packages/material-ui/src/Button/Button.js
@@ -18,6 +18,8 @@ export const styles = theme => ({
     minWidth: 64,
     minHeight: 36,
     padding: '6px 16px',
+    height: 0, // IE11 work around for alignItems
+    alignItems: 'center',
     borderRadius: theme.shape.borderRadius,
     color: theme.palette.text.primary,
     transition: theme.transitions.create(['background-color', 'box-shadow', 'border'], {

--- a/packages/material-ui/src/Button/Button.js
+++ b/packages/material-ui/src/Button/Button.js
@@ -18,7 +18,6 @@ export const styles = theme => ({
     minWidth: 64,
     minHeight: 36,
     padding: '6px 16px',
-    height: 0, // IE11 work around for alignItems
     alignItems: 'center',
     borderRadius: theme.shape.borderRadius,
     color: theme.palette.text.primary,
@@ -38,6 +37,10 @@ export const styles = theme => ({
     },
     '&$disabled': {
       color: theme.palette.action.disabled,
+    },
+    '&:before': {
+      content: '""',
+      height: 24, // Same height as desired min-height
     },
   },
   /* Styles applied to the span element that wraps the children. */
@@ -213,12 +216,20 @@ export const styles = theme => ({
     minWidth: 64,
     minHeight: 31,
     fontSize: theme.typography.pxToRem(13),
+    '&:before': {
+      content: '""',
+      height: 23, // Same height as desired min-height
+    },
   },
   /* Styles applied to the root element if `size="large"`. */
   sizeLarge: {
     padding: '8px 24px',
     minHeight: 42,
     fontSize: theme.typography.pxToRem(15),
+    '&:before': {
+      content: '""',
+      height: 26, // Same height as desired min-height
+    },
   },
   /* Styles applied to the root element if `fullWidth={true}`. */
   fullWidth: {

--- a/packages/material-ui/src/styles/createTypography.js
+++ b/packages/material-ui/src/styles/createTypography.js
@@ -71,7 +71,7 @@ export default function createTypography(palette, typography) {
     subtitle2: buildVariant(fontWeightMedium, 14, 1.57, 0.1),
     body1Next: buildVariant(fontWeightRegular, 16, 1.5, 0.15),
     body2Next: buildVariant(fontWeightRegular, 14, 1.5, 0.15),
-    buttonNext: buildVariant(fontWeightMedium, 14, 1.3125, 0.4, caseAllCaps),
+    buttonNext: buildVariant(fontWeightMedium, 14, 1.75, 0.4, caseAllCaps),
     captionNext: buildVariant(fontWeightRegular, 12, 1.66, 0.4),
     overline: buildVariant(fontWeightRegular, 12, 2.66, 1, caseAllCaps),
   };

